Cannot use injector.instanceOf() with @Cached

### Are you looking for help?

No

### Play Version (2.5.x / etc)

2.6.1

### API (Scala / Java / Neither / Both)

Java

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

Windows 7

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

### Library Dependencies

No

### Expected Behavior

Please refer the following code.
```
@Inject
play.inject.Injector injector;

@play.cache.Cached(key = "somewhat")
public Result index() {
  AnyClass o = injector.instanceOf(AnyClass.class);
  return ok();
}
```

I hope the instance of `AnyClass` will be retrived as expected.

### Actual Behavior

In fact, the injector throw an exception in DEV mode in sbt interactive mode.

```
! @74j450acf - Internal server error, for (GET) [/] ->

play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[CompletionException: java.lang.SecurityException: setContextClassLoader]]
        at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:255)
        at play.api.http.HttpErrorHandlerExceptions.throwableToUsefulException(HttpErrorHandler.scala)
        at play.http.DefaultHttpErrorHandler.throwableToUsefulException(DefaultHttpErrorHandler.java:204)
        at play.http.DefaultHttpErrorHandler.onServerError(DefaultHttpErrorHandler.java:164)
        at play.core.j.JavaHttpErrorHandlerAdapter.$anonfun$onServerError$1(JavaHttpErrorHandlerAdapter.scala:22)
Caused by: java.util.concurrent.CompletionException: java.lang.SecurityException: setContextClassLoader
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
        at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:961)
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926)
        at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
Caused by: java.lang.SecurityException: setContextClassLoader
        at java.util.concurrent.ForkJoinWorkerThread$InnocuousForkJoinWorkerThread.setContextClassLoader(ForkJoinWorkerThread.java:244)
        at play.api.inject.ContextClassLoaderInjector.withContext(Injector.scala:125)
        at play.api.inject.ContextClassLoaderInjector.instanceOf(Injector.scala:118)
        at play.inject.DelegateInjector.instanceOf(DelegateInjector.java:22)
        at controllers.Home.index(Home.java:49)
```

In PROD mode, the code works as expected.

### Reproducible Test Case

Just use the attached code as before.

This bit stood out to me:

```
Caused by: java.lang.SecurityException: setContextClassLoader
        at java.util.concurrent.ForkJoinWorkerThread$InnocuousForkJoinWorkerThread.setContextClassLoader(ForkJoinWorkerThread.java:244)
```

I did some research and it looks like this may be related to a change in the security behaviour of the ForkJoinPool in later versions of Java 8. Would you mind trying with an early release of Java 8 to see if that fixes the issue for you? It looks like the change occurred in build 122, so build 121 or earlier should be OK.

References:
- Someone with the same issue: https://stackoverflow.com/questions/21440563/webstart-application-securityexception-setcontextclassloader/21460363#21460363.
- Proposal for Java 8 change on mailing list: http://cs.oswego.edu/pipermail/concurrency-interest/2013-December/012123.html
- Patch to JDK: http://hg.openjdk.java.net/jdk8/jdk8/jdk/diff/e2bdddb8bedf/src/share/classes/java/util/concurrent/ForkJoinPool.java
- Summary of changes for build 122: http://mail.openjdk.java.net/pipermail/jdk8-dev/2013-December/003860.html

--- a/framework/src/play-cache/src/main/java/play/cache/CachedAction.java
+++ b/framework/src/play-cache/src/main/java/play/cache/CachedAction.java
@@ -3,7 +3,6 @@
  */
 package play.cache;
 
-import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 
 import play.mvc.Action;
@@ -27,14 +26,7 @@ public class CachedAction extends Action<Cached> {
     public CompletionStage<Result> call(Context ctx) {
         final String key = configuration.key();
         final Integer duration = configuration.duration();
-        return cacheApi.<Result>get(key).thenComposeAsync(cacheResult -> {
-            if (cacheResult != null) {
-                return CompletableFuture.completedFuture(cacheResult);
-            }
-            return delegate.call(ctx).thenComposeAsync(result ->
-                    cacheApi.set(key, result, duration).thenApply(ignore -> result)
-            );
-        });
+        return cacheApi.getOrElseUpdate(key, () -> delegate.call(ctx), duration);
     }
 
 }

--- a/framework/src/play-integration-test/src/test/scala/play/it/http/JavaActionCompositionSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/http/JavaActionCompositionSpec.scala
@@ -15,8 +15,6 @@ import play.mvc.{ EssentialFilter, Result, Results }
 import play.mvc.Http.Cookie
 import play.routing.{ Router => JRouter }
 
-import scala.collection.JavaConverters._
-
 class GuiceJavaActionCompositionSpec extends JavaActionCompositionSpec {
   override def makeRequest[T](controller: MockController, configuration: Map[String, AnyRef] = Map.empty)(block: WSResponse => T): T = {
     implicit val port = testServerPort
@@ -135,7 +133,7 @@ trait JavaActionCompositionSpec extends PlaySpecification with WsTestClient {
         Results.ok(request.username())
       }
     }) { response =>
-      val setCookie = response.allHeaders.get("Set-Cookie").mkString("\n")
+      val setCookie = response.headers.get("Set-Cookie").mkString("\n")
       setCookie must contain("PLAY_SESSION=; Max-Age=-86400")
       response.body must_== "foo"
     }
@@ -146,7 +144,7 @@ trait JavaActionCompositionSpec extends PlaySpecification with WsTestClient {
         Results.ok(request.username())
       }
     }) { response =>
-      val setCookie = response.allHeaders.get("Set-Cookie").mkString("\n")
+      val setCookie = response.headers.get("Set-Cookie").mkString("\n")
       setCookie must contain("PLAY_FLASH=; Max-Age=-86400")
       response.body must_== "foo"
     }
@@ -157,7 +155,7 @@ trait JavaActionCompositionSpec extends PlaySpecification with WsTestClient {
         Results.ok(request.username())
       }
     }) { response =>
-      val setCookie = response.allHeaders.get("Set-Cookie").mkString("\n")
+      val setCookie = response.headers.get("Set-Cookie").mkString("\n")
       setCookie must contain("foo=bar")
       response.body must_== "foo"
     }

--- a/framework/src/play-integration-test/src/test/scala/play/it/http/JavaCachedActionSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/http/JavaCachedActionSpec.scala
@@ -12,7 +12,6 @@ import com.google.common.primitives.Primitives
 import play.api.Application
 import play.api.cache.AsyncCacheApi
 import play.api.cache.ehcache.EhCacheModule
-import play.api.inject.{ ContextClassLoaderInjector, NewInstanceInjector, SimpleInjector }
 import play.api.inject.guice.GuiceApplicationBuilder
 import play.api.test.{ PlaySpecification, TestServer, WsTestClient }
 import play.cache.{ Cached, DefaultAsyncCacheApi }
@@ -78,13 +77,6 @@ class JavaCachedActionSpec extends PlaySpecification with WsTestClient {
         first.body must not(beEqualTo(second.body))
       }
 
-      "support using injector" in makeRequest(new CachedControllerWithInjector()) { port =>
-        val responses = BasicHttpClient.makeRequests(port)(
-          BasicRequest("GET", "/", "HTTP/1.1", Map(), "")
-        )
-
-        responses.head.status must beEqualTo(OK)
-      }
     }
 
     "when action is annotated" in {
@@ -132,29 +124,6 @@ class CachedController extends MockController {
     play.mvc.Results.ok("Cached result: " + System.currentTimeMillis())
   }
 }
-
-@Cached(key = "play.it.http.CachedControllerWithInjector.cache", duration = 1 /* second */ )
-class CachedControllerWithInjector extends MockController {
-  override def action: Result = {
-    injector.instanceOf(classOf[SomeComponent])
-    play.mvc.Results.ok("Cached result: " + System.currentTimeMillis())
-  }
-
-  def injector: play.inject.Injector = {
-    val thread = Thread.currentThread()
-    // injectception
-    new play.inject.DelegateInjector(
-      new ContextClassLoaderInjector(
-        new SimpleInjector(
-          fallback = NewInstanceInjector
-        ) + SomeComponent(), // register SomeComponent to SimpleInjector
-        thread.getContextClassLoader
-      )
-    )
-  }
-}
-
-case class SomeComponent()
 
 /**
  * This is necessary to avoid EhCache shutdown problems.

--- a/framework/src/play-integration-test/src/test/scala/play/it/http/JavaCachedActionSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/http/JavaCachedActionSpec.scala
@@ -3,7 +3,7 @@
  */
 package play.it.http
 
-import java.util.concurrent.{ CompletableFuture, TimeUnit }
+import java.util.concurrent.{ Callable, CompletableFuture, CompletionStage, TimeUnit }
 import javax.inject.{ Inject, Provider }
 
 import akka.Done
@@ -200,10 +200,12 @@ class TestAsyncCacheApiProvider @Inject() (lifeCycle: ApplicationLifecycle)(impl
       .expireAfterWrite(1, TimeUnit.SECONDS) // consistent with the value used in @Cached annotations above
       .build[String, Object]()
 
-    lifeCycle.addStopHook(() => {
-      cache.cleanUp()
-      cache.invalidateAll()
-      CompletableFuture.completedFuture(true)
+    lifeCycle.addStopHook(new Callable[CompletionStage[_]] {
+      override def call(): CompletionStage[_] = {
+        cache.cleanUp()
+        cache.invalidateAll()
+        CompletableFuture.completedFuture(true)
+      }
     })
 
     new TestAsyncCacheApi(cache)

--- a/framework/src/play-integration-test/src/test/scala/play/it/http/JavaCachedActionSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/http/JavaCachedActionSpec.scala
@@ -1,0 +1,211 @@
+/*
+ * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
+ */
+package play.it.http
+
+import java.util.concurrent.{ CompletableFuture, TimeUnit }
+import javax.inject.{ Inject, Provider }
+
+import akka.Done
+import com.github.benmanes.caffeine.cache.{ Cache, Caffeine }
+import com.google.common.primitives.Primitives
+import play.api.Application
+import play.api.cache.AsyncCacheApi
+import play.api.cache.ehcache.EhCacheModule
+import play.api.inject.{ ContextClassLoaderInjector, NewInstanceInjector, SimpleInjector }
+import play.api.inject.guice.GuiceApplicationBuilder
+import play.api.test.{ PlaySpecification, TestServer, WsTestClient }
+import play.cache.{ Cached, DefaultAsyncCacheApi }
+import play.inject.ApplicationLifecycle
+import play.mvc.Result
+
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.concurrent.duration._
+import scala.reflect.ClassTag
+
+class JavaCachedActionSpec extends PlaySpecification with WsTestClient {
+
+  def makeRequest[T](controller: MockController)(block: Port => T): T = {
+
+    import play.api.inject.bind
+
+    implicit val port = testServerPort
+    lazy val app: Application = GuiceApplicationBuilder()
+      .disable[EhCacheModule]
+      .bindings(
+        bind[play.api.cache.AsyncCacheApi].toProvider[TestAsyncCacheApiProvider],
+        bind[play.cache.AsyncCacheApi].to[DefaultAsyncCacheApi]
+      )
+      .routes {
+        case _ => JAction(app, controller)
+      }.build()
+
+    running(TestServer(port, app)) {
+      block(port)
+    }
+  }
+
+  "Java CachedAction" should {
+
+    "when controller is annotated" in {
+
+      "cache result" in makeRequest(new CachedController()) { port =>
+        val responses = BasicHttpClient.makeRequests(port)(
+          BasicRequest("GET", "/", "HTTP/1.1", Map(), ""),
+          BasicRequest("GET", "/", "HTTP/1.1", Map(), "")
+        )
+
+        val first = responses.head
+        val cached = responses.last
+
+        first.status must beEqualTo(cached.status)
+        first.body must beEqualTo(cached.body)
+      }
+
+      "expire result" in makeRequest(new CachedController()) { port =>
+
+        val first = BasicHttpClient.makeRequests(port)(
+          BasicRequest("GET", "/", "HTTP/1.1", Map(), "")
+        ).head
+
+        Thread.sleep(5.seconds.toMillis) // enough time to ensure the cache was expired
+
+        val second = BasicHttpClient.makeRequests(port)(
+          BasicRequest("GET", "/", "HTTP/1.1", Map(), "")
+        ).head
+
+        first.status must beEqualTo(second.status)
+        first.body must not(beEqualTo(second.body))
+      }
+
+      "support using injector" in makeRequest(new CachedControllerWithInjector()) { port =>
+        val responses = BasicHttpClient.makeRequests(port)(
+          BasicRequest("GET", "/", "HTTP/1.1", Map(), "")
+        )
+
+        responses.head.status must beEqualTo(OK)
+      }
+    }
+
+    "when action is annotated" in {
+      "cache result" in makeRequest(new MockController {
+        @Cached(key = "play.it.http.MockController.MockController.cache", duration = 1 /* second */ )
+        override def action: Result = play.mvc.Results.ok("Cached result: " + System.nanoTime())
+      }) { port =>
+        val responses = BasicHttpClient.makeRequests(port)(
+          BasicRequest("GET", "/", "HTTP/1.1", Map(), ""),
+          BasicRequest("GET", "/", "HTTP/1.1", Map(), "")
+        )
+
+        val first = responses.head
+        val cached = responses.last
+
+        first.status must beEqualTo(cached.status)
+        first.body must beEqualTo(cached.body)
+      }
+
+      "expire result" in makeRequest(new MockController {
+        @Cached(key = "play.it.http.MockController.MockController.cache", duration = 1 /* second */ )
+        override def action: Result = play.mvc.Results.ok("Cached result: " + System.nanoTime())
+      }) { port =>
+
+        val first = BasicHttpClient.makeRequests(port)(
+          BasicRequest("GET", "/", "HTTP/1.1", Map(), "")
+        ).head
+
+        Thread.sleep(5.seconds.toMillis) // enough time to ensure the cache was expired
+
+        val second = BasicHttpClient.makeRequests(port)(
+          BasicRequest("GET", "/", "HTTP/1.1", Map(), "")
+        ).head
+
+        first.status must beEqualTo(second.status)
+        first.body must not(beEqualTo(second.body))
+      }
+    }
+  }
+}
+
+@Cached(key = "play.it.http.CachedController.cache", duration = 1 /* second */ )
+class CachedController extends MockController {
+  override def action: Result = {
+    play.mvc.Results.ok("Cached result: " + System.currentTimeMillis())
+  }
+}
+
+@Cached(key = "play.it.http.CachedControllerWithInjector.cache", duration = 1 /* second */ )
+class CachedControllerWithInjector extends MockController {
+  override def action: Result = {
+    injector.instanceOf(classOf[SomeComponent])
+    play.mvc.Results.ok("Cached result: " + System.currentTimeMillis())
+  }
+
+  def injector: play.inject.Injector = {
+    val thread = Thread.currentThread()
+    // injectception
+    new play.inject.DelegateInjector(
+      new ContextClassLoaderInjector(
+        new SimpleInjector(
+          fallback = NewInstanceInjector
+        ) + SomeComponent(), // register SomeComponent to SimpleInjector
+        thread.getContextClassLoader
+      )
+    )
+  }
+}
+
+case class SomeComponent()
+
+/**
+ * This is necessary to avoid EhCache shutdown problems.
+ *
+ * Using Caffeine here since it is already a dependency and it handles expiration.
+ */
+class TestAsyncCacheApi(cache: Cache[String, Object])(implicit context: ExecutionContext) extends AsyncCacheApi {
+  override def set(key: String, value: Any, expiration: Duration): Future[Done] = Future.successful {
+    cache.put(key, value.asInstanceOf[Object])
+    Done
+  }
+
+  override def remove(key: String): Future[Done] = Future {
+    cache.invalidate(key)
+    Done
+  }
+
+  override def getOrElseUpdate[A: ClassTag](key: String, expiration: Duration)(orElse: => Future[A]): Future[A] = {
+    get[A](key).flatMap {
+      case Some(value) => Future.successful(value)
+      case None => orElse.flatMap(value => set(key, value, expiration).map(_ => value))
+    }
+  }
+
+  override def get[T](key: String)(implicit ct: ClassTag[T]): Future[Option[T]] = {
+    val result = Option(cache.getIfPresent(key)).filter { v =>
+      Primitives.wrap(ct.runtimeClass).isInstance(v) ||
+        ct == ClassTag.Nothing || (ct == ClassTag.Unit && v == ((): Unit))
+    }.asInstanceOf[Option[T]]
+    Future.successful(result)
+  }
+
+  override def removeAll(): Future[Done] = Future {
+    cache.invalidateAll()
+    Done
+  }
+}
+
+class TestAsyncCacheApiProvider @Inject() (lifeCycle: ApplicationLifecycle)(implicit context: ExecutionContext) extends Provider[TestAsyncCacheApi] {
+  override def get(): TestAsyncCacheApi = {
+    val cache = Caffeine
+      .newBuilder()
+      .expireAfterWrite(1, TimeUnit.SECONDS) // consistent with the value used in @Cached annotations above
+      .build[String, Object]()
+
+    lifeCycle.addStopHook(() => {
+      cache.cleanUp()
+      cache.invalidateAll()
+      CompletableFuture.completedFuture(true)
+    })
+
+    new TestAsyncCacheApi(cache)
+  }
+}

